src/libgit2/transports/emscriptenhttp.cc

```diff
@@ -0,0 +1,271 @@
+    if (f->status != 200) {
+      printf("%d %s %s\n", f->status, f->statusText, f->url);
+    }
+    if (f->readyState != 4) {
+      printf("Connection is not in ready state (%d)\n", f->readyState);
+    }
+  }
+  connection.writeBuffer.clear();
+  *bytesRead = min(f->numBytes - connection.totalBytesRead, bufferSize);
+  std::memcpy(buffer, f->data + connection.totalBytesRead, *bytesRead);
+  connection.totalBytesRead += *bytesRead;
+}
+
+void xhrFree(uint64_t connectionNumber) {
+  if (connectionMap.count(connectionNumber != 1)) {
+    printf("Attempting to free unkown connection %l", connectionNumber);
+    return;
+  }
+  auto &connection = connectionMap[connectionNumber];
+  emscripten_fetch_close(connection.fetch);
+  connectionMap.erase(connectionNumber);
+}
+}  // namespace
+
+namespace {
+typedef struct {
+  git_smart_subtransport_stream parent;
+  git_str service_url;
+  uint64_t connectionNo;
+} emscriptenhttp_stream;
+
+typedef struct {
+  git_smart_subtransport parent;
+  transport_smart *owner;
+} emscriptenhttp_subtransport;
+
+// Since these types are being sent back to C, we need to ensure they are PODs.
+static_assert(std::is_pod<emscriptenhttp_stream>());
+static_assert(std::is_pod<emscriptenhttp_subtransport>());
+
+static int emscriptenhttp_stream_read(
+  git_smart_subtransport_stream *stream, char *buffer, size_t buf_size, size_t *bytes_read) {
+  emscriptenhttp_stream *s = (emscriptenhttp_stream *)stream;
+
+  if (s->connectionNo == -1) {
+    s->connectionNo = xhrConnect(git_str_cstr(&s->service_url), "GET", {});
+  }
+  xhrRead(s->connectionNo, buffer, buf_size, bytes_read);
+
+  return 0;
+}
+
+static int emscriptenhttp_stream_write_single(
+  git_smart_subtransport_stream *stream, const char *buffer, size_t len) {
+  emscriptenhttp_stream *s = (emscriptenhttp_stream *)stream;
+
+  if (s->connectionNo == -1) {
+
+    auto serviceUrl = git_str_cstr(&s->service_url);
+    bool uploadPack = strstr(serviceUrl, "git-upload-pack") != 0;
+
+    s->connectionNo = xhrConnect(
+      serviceUrl,
+      "POST",
+      {
+        "Content-Type",
+        uploadPack ? "application/x-git-upload-pack-request"
+                   : "application/x-git-receive-pack-request",
+        "Pragma",
+        "no-cache",
+      });
+  }
+  xhrWrite(s->connectionNo, buffer, len);
+
+  return 0;
+}
+
+static void emscriptenhttp_stream_free(git_smart_subtransport_stream *stream) {
+  emscriptenhttp_stream *s = (emscriptenhttp_stream *)stream;
+  if (s->connectionNo != -1) {
+    xhrFree(s->connectionNo);
+  }
+  git_str_dispose(&s->service_url);
+  git__free(s);
+}
+
+static int emscriptenhttp_stream_alloc(
+  emscriptenhttp_subtransport *t, emscriptenhttp_stream **stream) {
+  emscriptenhttp_stream *s;
+
+  if (!stream)
+    return -1;
+
+  s = reinterpret_cast<emscriptenhttp_stream *>(git__calloc(1, sizeof(emscriptenhttp_stream)));
+  GIT_ERROR_CHECK_ALLOC(s);
+
+  s->parent.subtransport = &t->parent;
```

Comment:
Wait, is `&t->parent` the same as `&(t->parent)`? Pretty wild.

Reply:
Order of operations 🙃